Unit 3 : Vision Transformers / Transfer Learning & Fine-Tuning Chapter Content (#204)
johko/computer-vision-course: chapters/en/Unit 3 - Vision Transformers/Vision Transformers for Image Classification.mdx

### CNN vs Vision Transformers: Inductive Bias

Inductive bias is a term used in machine learning to describe the set of assumptions that a learning algorithm uses to make predictions. In simpler terms, inductive bias is like a shortcut that helps a machine learning model make educated guesses based on the information it has seen so far.

Here's a couple of inductive biases we observe in CNNs:

- Translational Equivariance: an object can appear anywhere in the image, and CNNs can detect its features.
- Locality: pixels in an image interact mainly with its surrounding pixels to form features.

These are lacking in Vision Transformers. Then how do they perform so well? It's because they're highly scalable and they're trained on massive amounts of images. Hence, they overcome the need for these inductive biases.

### Using pre-trained Vision Transformers

> Since it's not feasible for everyone to train a Vision Transformers on millions of images to get good performance. Instead we can take advantage of openly available models from places such as the [Hugging Face Hub](https://huggingface.co/models?sort=trending).

Reviewer: ```suggestion
It's not feasible for everyone to train a Vision Transformers on millions of images to get good performance. Instead, one can take advantage of openly available models from places such as the [Hugging Face Hub](https://huggingface.co/models?sort=trending).
```

> What to do with the pre-trained model? We can apply transfer learning and fine-tuning!

Reviewer: ```suggestion
What do you do with the pre-trained model? You can apply transfer learning and fine-tune it!
```

## Transfer Learning & Fine-Tuning for Image Classification

The idea of transfer learning is that we can leverage the features learned by the Vision Transformers trained on a very large dataset and apply these features to our dataset. This can lead to significant improvements in model performance, especially when our dataset has limited data available for training.

> Since we are taking advantage of the learned features, we do not need to update the entire model either. By freezing the majority of the weights, we can just choose to train only certain layers to get excellent performance with less training time and low GPU consumption.

Reviewer: ```suggestion
Since we are taking advantage of the learned features, we do not need to update the entire model either. By freezing the majority of the weights, we can choose to train only certain layers to get excellent performance with less training time and low GPU consumption.
```

### Multi-class Image Classification

You can go through the transfer learning tutorial using Vision Transformers for image classification in this notebook:

<a
  target="_blank"
  href="https://colab.research.google.com/github/johko/computer-vision-course/blob/main/notebooks/Unit%203%20-%20Vision%20Transformers/transfer-learning-image-classification.ipynb"
>
  <img
    src="https://colab.research.google.com/assets/colab-badge.svg"
    alt="Open In Colab"
  />
</a>

> This is what we'd be building, an image classifier to classify dog and cat breeds:

Reviewer: ```suggestion
This is what we'll be building: an image classifier to tell apart dog and cat breeds:
```

<Tip>
  However, in most scenarios, applying transfer learning is ample in the case of
  Vision Transformers.
</Tip>

### Multi-label Image Classification

The tutorial above teaches multi-class image classification, where each image only has 1 class assigned to it. What about scenarios where each image has multiple labels in a multi-class dataset?

This notebook will walk you through a fine-tuning tutorial using Vision Transformer for multi-label image classification:

<a
  target="_blank"
  href="https://colab.research.google.com/github/johko/computer-vision-course/blob/main/notebooks/Unit%203%20-%20Vision%20Transformers/fine-tuning-multilabel-image-classification.ipynb"
>
  <img
    src="https://colab.research.google.com/assets/colab-badge.svg"
    alt="Open In Colab"
  />
</a>

We'll also be learning how to use [Hugging Face Accelerate](https://huggingface.co/docs/accelerate/index) to write our custom training loops.

> This is the outcome of the multi-label classification tutorial:

Reviewer: ```suggestion
This is what you can expect to see as the outcome of the multi-label classification tutorial:
```
